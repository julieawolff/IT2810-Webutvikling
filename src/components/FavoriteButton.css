.favoriteButton {
  max-height: 50px;
  max-width: 55px;
  border: transparent;
  padding: 0.4em;
  font-size: 1.5em;
  font-family: system-ui;
  background-color: transparent;
  cursor: pointer;
}
