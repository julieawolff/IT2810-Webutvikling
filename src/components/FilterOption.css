.filterOption {
  background-color: #ffffff;
  border: 3px solid transparent;
  padding: 10px;
  border-radius: 8px;
  display: flex;
  align-items: start;
  margin-bottom: 10px;
}

.filterOptionLabel {
  font-size: 16px;
  color: #333;
  margin-left: 5px;
}

.filterOption:hover {
  font-weight: bold;
}
