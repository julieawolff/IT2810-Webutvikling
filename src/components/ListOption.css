.listOption {
  background-color: #ffffff;
  border: 3px solid transparent;
  padding: 10px;
  border-radius: 10px;
  display: flex;
  align-items: center;
  margin-bottom: 10px;
  cursor: pointer;
  width: 100%;
}

.listOptionLabel {
  font-size: 16px;
  color: #333;
}

.listOption:hover {
  font-weight: bold;
}
