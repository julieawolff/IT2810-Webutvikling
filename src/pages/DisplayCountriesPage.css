.outerDiv {
  display: flex;
  flex-direction: row;
  align-content: center;
  width: 100vw;
  height: 92vh;
  background-color: #dfe4ff;
}

@media (max-width: 768px) {
  .outerDiv {
    flex-direction: column;
    height: 98vh;
    width: 100vw;
  }
}

@media (max-height: 530px) {
  .outerDiv {
    height: fit-content;
  }
}
