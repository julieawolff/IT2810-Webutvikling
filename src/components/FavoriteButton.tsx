import { useEffect, useState } from 'react';
import './FavoriteButton.css';

interface FavoriteButtonProps {
  initialFavoriteStatus: boolean;
  handleFavorite: () => void;
}

export default function FavoriteButton({ initialFavoriteStatus, handleFavorite }: FavoriteButtonProps) {
  const [isFavorite, setIsFavorite] = useState(initialFavoriteStatus);

  useEffect(() => {
    setIsFavorite(initialFavoriteStatus);
  }, [initialFavoriteStatus]);

  const handleFavoriteClick = () => {
    setIsFavorite(!isFavorite);
    handleFavorite();
  };

  return (
    <button className="favoriteButton" onClick={handleFavoriteClick}>
      {isFavorite ? '\u2665' : '\u2661'}
    </button>
  );
}
