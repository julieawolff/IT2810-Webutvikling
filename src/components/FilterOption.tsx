import { useState } from 'react';
import './FilterOption.css';

interface FilterOptionProps {
  checked: boolean;
  label: string;
  onChange: (checked: boolean) => void;
}

export default function FilterOption({ label, checked, onChange }: FilterOptionProps) {
  const [isChecked, setIsChecked] = useState(checked);

  const handleCheckboxChange = () => {
    setIsChecked(!isChecked);
    onChange(!isChecked);
  };

  return (
    <label className="filterOption">
      <input
        className="filterOptionCheckbox"
        type="checkbox"
        checked={isChecked}
        onChange={handleCheckboxChange}
      />
      <span className="filterOptionLabel">{label}</span>
    </label>
  );
}
