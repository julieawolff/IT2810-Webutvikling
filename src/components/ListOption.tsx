import './ListOption.css';

interface ListOptionProps {
  changedCountryName: string;
  changeCountry: (changedCountryName: string) => void;
}

export default function ListOption({ changedCountryName, changeCountry }: ListOptionProps) {
  return (
    <button className="listOption" onClick={() => changeCountry(changedCountryName)}>
      <span className="listOptionLabel">{changedCountryName}</span>
    </button>
  );
}
