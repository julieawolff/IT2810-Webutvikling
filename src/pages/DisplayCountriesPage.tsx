import { useState, useEffect } from 'react';
import CountryCard from '../components/CountryCard';
import './DisplayCountriesPage.css';
import { useQuery } from '@tanstack/react-query';
import FilterOptionOverview from '../components/FilterOptionOverview';

//Henter alle landene fra REST Countries API-et
const fetchCountries = async () => {
  const response = await fetch(
    'https://restcountries.com/v3.1/all?fields=name,continents,capital,flags,languages,currencies',
  );
  const data = await response.json();
  return data.slice(0, 60); //Kan endre her for å endre hvilke/hvor mange land vi henter ut
};

export default function DisplayCountriesPage() {
  const { data, error, isLoading } = useQuery({ queryKey: ['countries'], queryFn: fetchCountries });
  const [currentCountryIndex, setCurrentCountryIndex] = useState(0);

  const [selectedOptions, setSelectedOptions] = useState<string[]>([]);

  const [favoriteCountries, setFavoriteCountries] = useState<string[]>([]);

  //Henter de favorittlandene som er lagret i local storage
  useEffect(() => {
    try {
      const storedFavorites = JSON.parse(localStorage.getItem('favoriteCountries') || '[]');
      setFavoriteCountries(storedFavorites);
    } catch (error) {
      console.error('Failed to parse favorite countries from local storage:', error);
      setFavoriteCountries([]);
    }
  }, []);

  //Henter de filtreringsvalgene som er lagret i session storage
  useEffect(() => {
    try {
      const storedSelectedOptions = JSON.parse(sessionStorage.getItem('selectedOptions') || '[]');
      setSelectedOptions(storedSelectedOptions);
    } catch (error) {
      console.error('Failed to parse selected options from session storage:', error);
      setSelectedOptions([]);
    }
  }, []);

  if (isLoading) return <p>Laster...</p>;
  if (error) return <p>En feil oppsto: {error.message}</p>;

  //Alle landene som hentes fra REST countries
  const allCountryData = data;

  // Dymaisk liste som endrer seg basert på hva som filtrers på (dersom ingen filtreringer er gitt har man bare hele allCountryData)
  let currentCountries = [];
  if (selectedOptions.length === 0) {
    currentCountries = allCountryData;
  } else if (selectedOptions.includes('Favorites') && favoriteCountries.length === 0 && selectedOptions.length === 1) {
    currentCountries = allCountryData;
  } else {
    currentCountries = allCountryData.filter(
      (country: { continents: string[]; name: { common: string } }) =>
        selectedOptions.includes(country.continents[0]) ||
        (selectedOptions.includes('Favorites') && favoriteCountries.includes(country.name.common)),
    );
  }

  //Landet som skal vises i CardComponent
  const currentCountry = currentCountries[currentCountryIndex];

  //Formatering på dataen som blir hentet via API-et
  const continent = currentCountry.continents[0] ?? 'None';

  const name = currentCountry.name.common ?? 'None';

  const capital = currentCountry.capital ?? 'None';

  const flag = currentCountry.flags.png ?? 'None';

  //Languages og currencies hentes som dictionaries
  //Antartica har eksempelvis verken capital, languages eller currencies og disse vil derfor bli hentet ut som tomme objekt aka ha lengde 0
  const languages =
    currentCountry.languages && Object.keys(currentCountry.languages).length > 0
      ? Object.values(currentCountry.languages).slice(0, 2).join(', ')
      : 'None';

  const currencies =
    currentCountry.currencies && Object.keys(currentCountry.currencies).length > 0
      ? (Object.values(currentCountry.currencies)[0] as { name: string })?.name
      : 'None';

  //To funksjoner for hva som skal skje dersom hendholsvis pil venstre og pil høyre trykkes på
  //Kalles på fra CountryCardComponent
  const handleLeft = () => {
    if (currentCountryIndex == 0) {
      setCurrentCountryIndex(currentCountries.length - 1);
    } else {
      setCurrentCountryIndex(currentCountryIndex - 1);
    }
  };

  const handleRight = () => {
    if (currentCountryIndex == currentCountries.length - 1) {
      setCurrentCountryIndex(0);
    } else {
      setCurrentCountryIndex(currentCountryIndex + 1);
    }
  };

  //Funskjon som håndterer endring av valgte kontinenter
  const handleOptionChange = (currentOption: string, checked: boolean) => {
    let updatedSelectedOptions: string[] = [];

    if (checked) {
      updatedSelectedOptions = [...selectedOptions, currentOption];
    } else {
      updatedSelectedOptions = selectedOptions.filter((option) => option !== currentOption);
    }
    setSelectedOptions(updatedSelectedOptions);
    sessionStorage.setItem('selectedOptions', JSON.stringify(updatedSelectedOptions));
    setCurrentCountryIndex(0);
  };

  // Kalles når brukeren trykker på favorittknappen -> Oppdaterer favoriteCountries og lagrer i local storage
  const handleFavorite = (countryName: string) => {
    let updatedFavorites: string[] = [];
    if (favoriteCountries.includes(countryName)) {
      updatedFavorites = favoriteCountries.filter((country) => country !== countryName);
      window.location.reload();
    } else {
      updatedFavorites = [...favoriteCountries, countryName];
    }
    setFavoriteCountries(updatedFavorites);
    localStorage.setItem('favoriteCountries', JSON.stringify(updatedFavorites));
  };

  const handleChangeCountry = (changedCountryName: string) => {
    //Finne index til landet vi ønsker å vise i CountryCardComponent
    const countryIndex = currentCountries.findIndex(
      (country: { name: { common: string } }) => country.name.common === changedCountryName,
    );
    //Sette CurrentCountryIndex til navnet på den ListOption-en som er blitt trykket på
    if (countryIndex !== -1) {
      setCurrentCountryIndex(countryIndex);
    } else {
      console.log(`Country with the name "${name}" not found.`);
    }
  };

  return (
    <div className="outerDiv">
      <FilterOptionOverview
        changeCountry={handleChangeCountry}
        handleOptionChange={handleOptionChange}
        selectedOptions={selectedOptions}
        currentContries={currentCountries.map((country: { name: { common: string } }) => country.name.common)}
      />
      <CountryCard
        country={name}
        continent={continent}
        currency={currencies}
        capital={capital}
        language={languages}
        flag={flag}
        onLeft={handleLeft}
        onRight={handleRight}
        handleFavorite={() => handleFavorite(currentCountry.name.common)}
        isCountryFavorite={favoriteCountries.includes(currentCountry.name.common)}
      ></CountryCard>
    </div>
  );
}
