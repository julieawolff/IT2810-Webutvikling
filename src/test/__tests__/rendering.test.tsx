import { render, screen, waitFor } from '@testing-library/react';
import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import FilterOption from '../../components/FilterOption';
import FavoriteButton from '../../components/FavoriteButton';
import DisplayCountriesPage from '../../pages/DisplayCountriesPage';
import ListOption from '../../components/ListOption';

it('renders DisplayCountriesPage with mock data', async () => {
  const queryClient = new QueryClient();
  render(
    <QueryClientProvider client={queryClient}>
      <DisplayCountriesPage />
    </QueryClientProvider>,
  );

  await waitFor(() => {
    expect(screen.getAllByText('Norway')[0]);
    expect(screen.getAllByText('Capital: Oslo')[0]);
    expect(screen.getAllByText('Language: Norwegian')[0]);
  });
});

describe('FavoriteButton Snapshot Tests', () => {
  it('renders correctly when not a favorite', () => {
    const { asFragment } = render(<FavoriteButton initialFavoriteStatus={false} handleFavorite={() => {}} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders correctly when it is a favorite', () => {
    const { asFragment } = render(<FavoriteButton initialFavoriteStatus={true} handleFavorite={() => {}} />);
    expect(asFragment()).toMatchSnapshot();
  });
});

describe('FilterOption Component', () => {
  it('renders with correct label and checked status', () => {
    render(<FilterOption label="Test Option" checked={false} onChange={() => {}} />);

    
    expect(screen.getByText('Test Option'));

    const checkboxElement = screen.getByRole('checkbox') as HTMLInputElement;
    expect(checkboxElement.checked).toBe(false);
  });
});

describe('ListOption Component', () => {
  it('renders with correct name', () => {
    render(<ListOption changedCountryName="Fiji" changeCountry={() => {}} />);
    expect(screen.getByText('Fiji'));
  });
});
