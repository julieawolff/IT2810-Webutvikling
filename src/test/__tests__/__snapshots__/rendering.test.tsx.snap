// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`FavoriteButton Snapshot Tests > renders correctly when it is a favorite 1`] = `
<DocumentFragment>
  <button
    class="favoriteButton"
  >
    ♥
  </button>
</DocumentFragment>
`;

exports[`FavoriteButton Snapshot Tests > renders correctly when not a favorite 1`] = `
<DocumentFragment>
  <button
    class="favoriteButton"
  >
    ♡
  </button>
</DocumentFragment>
`;
